### DISCUSSION TITLE
Raspberry Pi 5

### DISCUSSION THREAD

**Discussion by spikked27** [awawa-dev/HyperHDR#661] [Q&A] [answered]
Just a general question if there is any planned support for the raspberry pi 5 now that units are being shipped.  Thanks!

**Comment by awawa-dev** [comment]
My RPi 5 arrives next week so let's see. I don't expect any problems with HyperHDR itself. However, as usual, there will probably be problems with the external rpi_ws281x, but for me this is a secondary issue because it has not been officially supported here for a long time. And now, in addition, the chipset responsible for GPIO has completely changed and it is not known what the consequences will be, I expect SPI should work 1:1 because it's a core & popular protocol and I can't imagine that they didn't test it before releasing Raspberry Pi 5. And for PWM who knows... I warned about it, so users using this driver chose themselves their fate. But there are alternatives or they have to wait for possible changes in the upstream of this library. There are also serious changes in the latest OS based on Bookworm, WiFi configuration will no longer be so simple and requires the use of the Raspberry Pi Imager ... but these are all peripherals around the application, HyperHDR itself should not require any changes.

**Comment by spikked27** [comment]
Awesome!  I just picked up a Pi 4 at Microcenter and realize I don't have anything to use it for, only have two Pis in my house, 1 running HyperHDR and the other Home Assistant (they've said support may not come until 2024), personally I use HyperSPI for my lights so looking forward to a HyperHDR release so I can see what kind of performance gains there are!  (Even if it's just the whole thing running a bit cooler)

**Comment by awawa-dev** [comment]
On my Rpi4 and Elgato I have a video frame decompression time around ~1ms, so I doubt that RPi5 will make a revolution here ;) But code compilation will be much faster on RPi5. It is not at all certain whether the RPi 5 will be cooler, since it now has a dedicated active cooling with dedicated holes for it on the PCB.

**Reply by spikked27** [reply]
Ok, you got me, I'll just have piece of mind knowing I'm running the latest hardware ;)

**Reply by awawa-dev** [reply]
On the other hand RPi 5 RTC it's a very nice feature :) A sudden jump in system time of up to several months forward or backward (for OS on  read-only drive) after RPi startup when NTP synchronization is finalized, brings unexpected challenges to the software.

**Comment by awawa-dev** [comment, marked as answer]
I had a little fun with RPi5, generally as I expected:

1) SPI seems to work after turning on (/boot/config.txt). I just need to check the signal with a logic analyzer to be 100% sure.  There is a new interface on /dev/spidev10.0 but we are not touching it! because it is used for something completely different.
2) **I checked and rpi_ws281x library doesn't work**... and it's not even about the new chipset revision number, which can be added to the library sources using notepad and recompiling it, but now the base address of the bcm2712 peripherals is shared with the new RP1 chipset (everything used to be in one continues space of single BCM in the past). It probably requires more work but I don't know the specifics of this project...maybe there is another solution?
**Use recommended HyperSPI, HyperSerialESP32 or HyperSerialPico instead.**
3) Rpi5 is extremely fast when compiling sources
4) not necessarily a completely scientific approach, just a quick test when capturing video using 1080p120 nv12 format (without any frame size or framerate reduction) in HyperHDR:
- RPi4 (bullseye OS) with Elgato HD60X: frame decoding time 18ms, CPU load 150% (out of 400%)
- RPi5 (bookworm OS) with Ezcap320: frame decoding time **4ms**, CPU load 95% (out of 400%) **Excellent score!** BUT read further...
5) Rpi5 got terribly hot during the 1080p120 nv12 test, RPi4 was much cooler. The RPi 5 exceeded 80C almost immediately, throttling around 85C stopped any further increase, while RPi4 even after a long time remained around 75C. Active cooling is rather mandatory. If we ignore the temperature, the RPi5 is a beast.
6) one standard plastic Raspberry Pi 3 case + one Raspberry Pi 4 case = one Raspberry Pi 5 case ..well, almost since there is no jack audio anymore ;)

**Reply by spikked27** [reply]
Great news!  I kinda figured the active cooler would be a necessity, it runs pretty quiet so not a huge deal for my use case at least.  Looking forward to trying out a build, I run an MS2130 with HyperSPI controlling 2 segments of 634 LEDs total, have it running at 720p 60Hz from the grabber and 60Hz on the LEDs but have to use Quarter Frame Mode to prevent lagging.

**Reply by awawa-dev** [reply]
'Quarter of frame' does not break anything but it is not necessary even for RPi4 and MS2130 * 2x300LEDs dummy devices despite I'm using the development version:
<details>
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/05dcdadc-a4c4-47e0-8f94-b8df653bbbb8)
</details>
Raw capturing without the 'quarter of frame' option enabled. Instead of the quarter mode you can also select 640x480x60 YUYV for the grabber (preferred way).

**Reply by spikked27** [reply]
My mistake, I'm capturing at 1280x720 not 720p hence my need for quarter frame mode, with 144 LEDs/m I'm liking the higher resolution capturing and with quarter frame mode the response is FAST

**Reply by awawa-dev** [reply]
Isn't 720p another word of 1280x720? Anyway it is better to use a lower resolution than the higher one and divide its width and size by 2 later (a quarter of the frame mode), because a larger frame must be sent over the bus and later saved in memory and this causes additional costs/load.

**Reply by jingato** [reply]
So if putting a new system together, is there currently any down side (besides the heat) to using a 5 over a 4? Any benefit to sticking with a 4? Thanks!

**Comment by awawa-dev** [comment]
I confirm that the Raspberry Pi 5 SPI interface works flawlessly. Below is a photo from the HyperSPI tests.
Yes, finally HyperSPI will support Raspberry Pi Pico (rp2040) also. The work is almost finished.

![hyperspi_pico_3](https://github.com/awawa-dev/HyperHDR/assets/69086569/ded972d5-1abb-40ad-8542-f136539be365)


As can be seen, I already have a case with a heat sink equipped with a fan and a SSD connected via USB3. Under heavy load, the temperature never exceeds 70C and the fan starts up, but it is almost silent. In idle and with a small load, the temperature is around 40-45C and the fan is then turned off.

More benchmarks:

Measured compilation time of the HyperHDR development version for the default configuration

`make clean && cmake -DCMAKE_BUILD_TYPE=Release .. && time make -j 4`

```
Raspberry Pi 5 (4GB, USB3.0 NVME PCI-E, Bookworm OS)
real    4m38.312s
user    15m37.099s
sys     0m45.744s

Raspberry Pi 4 (8GB, USB3.0 NVME PCI-E, Bullseye OS)
real    10m21.641s
user    34m41.694s
sys     3m9.223s
```

#raspberrypi5 #benchmark #rpi5

**Comment by awawa-dev** [comment]
@jingato 
> So if putting a new system together, is there currently any down side (besides the heat) to using a 5 over a 4? Any benefit to sticking with a 4? Thanks!

RPi5:

- mentioned problem with overheating... but it has a cool connector for a fan (sold separately)
- requires a Raspberry Pi system based on Bookworm, whose stability/quality is questionable at least for now. The older stable version of bullseye will not boot on RPi5.
- GPIO management, power supply and distribution (e.g. the 5V pin is no longer connected directly to the USB power line) have been completely changed
- new PMIC, past experience recommends caution and using original power supply (sold separately)
- lack of diagram and full documentation (it seems that the introduction was a bit hasty)
- some peripherals are not currently compatible with Rpi5 and some will never be
- no audio jack or analog audio, but has RTC if you need it (battery sold separately)
- no PWM control for LEDs if someone still tries to use it (in the future PIO may be an interesting alternative, but again: no documentation)
- and this is priceless: the 8GB version is slower than the 4GB version. Choose wisely how much RAM you really need.

Summary: I have one RPi5 4GB from my first pre-order, but I canceled the RPi5 8GB and all other peripherals.

**Reply by jingato** [reply]
Thanks. Sounds like I should stick with the 4 for now. I don't plan on using it for anything but this project.

**Comment by Kartel1** [comment]
Hello,
I have a RPi5 and I try to control an apa102 led strip with the GPIO. I want to do some test before cutting and soldering the strip.
However, I only have the first led of the strip light up the other are not and nothing change when I try to control the strip.
I'm using a hat which working as a level shifter to convert 3V to 5v.
Can you help me to understand what I am doing wrong please?

**Reply by awawa-dev** [reply]
Hi
What type of level shifter is it? Maybe that's the problem, because eg. I2C level shifters usually wont work. Does the first LED change colors or it's static? There is also a risk that the connector is broken.

**Reply by Kartel1** [reply]
Thx for the fast reply
Actually I have tried without the hat and I had no control over the strip, that's why I've order the level shifter.
The first led is static, but I think I can notice some intensity shift, but the color remain the same. 
The only time I was able to control this led strip was when I was using an Rpi4 with some jumper. I didn't try the jumper on rpi5 as I don't have jumper anymore.
The current level shifter I'm using is this one: https://www.joy-it.net/en/products/RB-LLC.

When you say there is a risk the connector is broken, which connector do you mean?

**Reply by awawa-dev** [reply]
If Rpi4 worked then the connector (on the LED strip side) is probably fine. Are you sure you selected correct SPI interface (RPI5 has additional one for internal purpose)? How did you manage to connect wires to pinout without jumpers? The level shifter looks like I2C type but I'm not 100% sure.

**Reply by Kartel1** [reply]
I was able to control led on rpi 4 with both hyperhdr and a python library, on hyperhdr I was using the standard spi (dev0.0). For the test on pi 4 I used jumper straight to the gpio.
For the test with rpi5 I use some hat with connectors like this one : https://fr.aliexpress.com/item/1005005012945817.html?spm=a2g0n.order_detail.order_detail_item.3.f7417d569HoiOR&gatewayAdapt=glo2fra

I will try with jumper straight connection as well, I need to get some first.

Right now if nothing is working on rpi5 I'm considering to buy an rpi4 or a wemos mini, because I'm pissed hahaha
(BTW this is my second ambilight, first one with Hyperion)

**Reply by awawa-dev** [reply]
SPI on RPi5 is a different type than on previous Raspberry Pi boards. Instead of Broadcom they are using their own RP1 chipset. But usually it should work (eg. HyperSPI works fine on RPi5). There is simply test https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#troubleshooting-spi-hardware but you should adjust the speed to the one you are using. No errors in HyperHDR logs? Anyway esp32 or Pico (rp2040, even better if with a level shifter) is also a good solution.

**Reply by Kartel1** [reply]
Ok then I will try it with the straight connection jumper.
Actually I have an error in the log: **_The refresh timer is enabled ('Refresh time' > 0) and may limit the performance of the LED driver. Ignore this error if you set it on purpose for some_** and i didn't know what to do about that because this value is locked on properties :D
I know it's better to use and esp + level shifter, but I'm a lazy guy and I wanted to avoid extra complexity and soldering as I can connected the led strip right away on the rpi (at least for the rpi4)  hahaha

**Reply by Kartel1** [reply]
Also, I took a look on the SPI test, but i'm not sure how i can conclude the spi is faulty or not.
And also I have a 2nd rpi5 so I can do some extra test if the first one might be faulty

**Reply by awawa-dev** [reply]
> Actually I have an error in the log: The refresh timer is enabled ('Refresh time' > 0) and may limit the performance of the LED driver. Ignore this error if you set it on purpose for some and i didn't know what to do about that because this value is locked on properties :D

Probably you set it at the beginning. Disable smoothing for a moment and set it in the LED driver to zero. Then enable smoothing again. Anyway it's ignored if the smoothing is enabled.

**Reply by Kartel1** [reply]
I don't think I have set this value.
I will try but i remember I've disabled smoothing with not success

**Reply by awawa-dev** [reply]
Maybe you need to refresh the page after you disabled it

**Reply by Kartel1** [reply]
I'm still not able to change the refresh time in led Hardware.
But without smoothing I have this in the log : _**2024-03-23T14:24:29.731Z [PERFORMANCE] [INSTANCE0: FPS = 5.50, processed = 330], [LED0: FPS = 32.42, send = 1945, processed = 326, dropped = 1053]**_
My test is to control led with full color mood blobs animation

**Reply by awawa-dev** [reply]
In case of SPI such logs dont say that the LED strip received the frames. Probably it appeared now because you waited bit longer (they are update each minute).

**Reply by Kartel1** [reply]
I don't understand what can make the app logging as it looks like the led doesn't receive the datas.
I'm confuse right now haha

**Reply by awawa-dev** [reply]
SPI communication is stateless. You can disconnect SPI to LED cables from RPi and it wont cause that some errors suddenly appear.

**Reply by Kartel1** [reply]
But an event that cause the error should occurs in order to have some log, I am a developer and that's why I'm confuse when you're telling me this

**Reply by awawa-dev** [reply]
it's a one way communication without any confirmation of reception. The error may be caused by using wrong (not existing) SPI interface but not by a such thing as no SPI device (like LED) attached to SPI interface. We only know that we sent frames, not that they were received.

**Reply by Kartel1** [reply]
Ah OK, that's much more clear to me now, thank you.
But the weird part is the drop. That's meaning their is some data that has been lost. How can that be logged if we can't check that frames has been received? 
I will do couple more test, especially with jumper straight connection to the rpi5 and I will see what I'm gonna do.

**Reply by awawa-dev** [reply]
Let it work for a few minutes after the start and upload full logs then

**Reply by Kartel1** [reply]
Sooooo!
I was able to do many tests.

1. I test the APA102 with another rpi5 with the 2 hats I used (without the level shifter hat) and the strip remains dark
2. I test the APA102 with jumper straight on rpi5 GPIO (no level shifter) and the strip remains dark
3. I test with an old led strip (WS2801) with jumper straight on rpi5 (no level shifter) and I was able to control the strip (I had some led at the end dark because of poor soldering)
4. I test the WS2801 with the 2 hats (still no level shifter) only 1 of the hat let me have control on the WS2801 (the other one was already suspected to be faulty)

The conclusion is have burnt my APA102 LED I guess, because it's working with a WS2801.
The other point is, I don't feel to use jumper for the connection for the final set up. I mean it works, but don't feel it's safe or polished.
I will have to figure out how i'm gonna finish the set up to have something clean.
Anyway now I have to order a new led strip.
Last thing, I've updated hyperhdr to the latest version and multiple times when I modify params in the UI, the server instance crash. I'm on version 20.0.0.0, is it LTS?

Still thanks a lot for helping me and working on this great project

**Reply by awawa-dev** [reply]
Thanks. I remember few cases that the first LED on the tape was damaged, also if you have a multimeter check if you have a short circuit on the APA LED strip between +5,data,clock,ground.
Also ws2801 is using lower speed than APA102. Maybe it's related to your problem?

> Last thing, I've updated hyperhdr to the latest version and multiple times when I modify params in the UI, the server instance crash. I'm on version 20.0.0.0, is it LTS?

need more information about what parameters you modified when it crashed: a path to reproduce the problem.

**Reply by Kartel1** [reply]
I don't think it's a shortcut as the apa was working on rpi4.
When you are talking about lower speed do you refer to the baud rate? Because it was set higher for the ws2801.
But do you think I should buy another apa102 or should I order another type? 
It has to operate on 5V because I don't want to change psu 

And for the instance crash it was related to changes on the led hardware section ( the baud rate for instance). But maybe it's related only to my install, can't tell

**Reply by awawa-dev** [reply]
Can you test APA102 again on Rpi4? The connector may have experienced mechanical damage. What speed were you using for APA102? For example Neopixelbus is using 10MHz on default.

> And for the instance crash it was related to changes on the led hardware section ( the baud rate for instance). But maybe it's related only to my install, can't tell

Thanks. I will check it later.

**Reply by awawa-dev** [reply]
> But do you think I should buy another apa102 or should I order another type?

In my setups I'm using only sk6812 RGBW cold white because it offers best color reproduction in my opinion.

**Reply by Kartel1** [reply]
I don't have a rpi4 anymore, I will have to ask a colleague if he can lend me one.
Maybe I should consider to test again on a rpi4 before ordering a new strip.

Actually I've played with the baud rate on apa102 from 1Mhz up to 10 Mhz. And if I remember correctly it was working with 10Mhz on the rpi4.

BTW with the latest version of hyperhdr it's set on 5Mhz for apa102 and I think I had crash when I tried to put it back to 10Mhz

**Reply by awawa-dev** [reply]
Since it doesn't work anyway maybe remove first LED and solder cables to the second one.

**Reply by Kartel1** [reply]
I wanted to do so, but I saw 2 microships ( for CI and Di I guess) before the 1st led.
Is it OK to use the led strip without that?
No safety issue or technical issue?

**Reply by Kartel1** [reply]
BTW I only have 16 awg cable, so that will be sport to solder with that on apa102 hahaha

**Reply by awawa-dev** [reply]
APA102 can be cut off on the marking lines. You cant use the old connector cables?

**Reply by Kartel1** [reply]
I mean I know I can cut it off on the mark, but I was wondering if it won't be any issue using the strip without the 2 microships. But it doesn't work currently so Yolo I guess hahaha.
I can use the old connector but they are kinda short and also since my psu is delivering 26A I would prefer not to use 22 awg. Maybe I should consider using some 18 awg idk

**Reply by awawa-dev** [reply]
Probably these are just resistors to limit current just in case. Anyway it's your choice. Of course, you can use even very thick cables, but remember that the current then flows through very thin copper paths along the LED strip. So without voltage injections in several places, it may be pointless.

**Reply by Kartel1** [reply]
Actually I saw the 2 same microships every 40 or 50 cm on the strip.
Anyway I think I will try that. Because I need to save money hahaha 

For the power supply I'm supplying the strip from both ends to have enough current. The thing is as I'm using a 26 A because I need at least 18A for the strip and the plan was to supply the rpi with the same psu. And for that kind of Amp I've read it's safer with 16awg that's why I choose that kind of wire.
But many ppl told me 18 awg is enough. And it's thinner than 16 awg

**Reply by awawa-dev** [reply]
> And for the instance crash it was related to changes on the led hardware section ( the baud rate for instance). But maybe it's related only to my install, can't tell

Yeah, I managed to reproduce it on my system. Thank you. I guess I shouldn't trust that much Github test phase here. Anyway I'm going to take a long vacation now so there will be plenty of time for testing then. 
Updated version will be ready in few hours. Please reinstall it later.

**Reply by Kartel1** [reply]
That's great you were able to find it out, I know debugging can be horrible sometimes hahaha.
Thx for the update, I will try it later.
Enjoy your vacation and thanks again a lot for your help and your time
